vispy/visuals/surface_plot.py

```diff
@@ -94,64 +77,106 @@ def set_data(self, x=None, y=None, z=None, colors=None):
                     self._z.shape != self.__vertices.shape[:2]):
                 self.__vertices = None
 
-        if self._z is None:
-            return
-
-        update_mesh = False
+    def _update_mesh_vertices(self, x_is_new, y_is_new, z_is_new):
         new_vertices = False
+        update_vertices = False
+        update_faces = False
 
         # Generate vertex and face array
         if self.__vertices is None:
-            new_vertices = True
             self.__vertices = np.empty((self._z.shape[0], self._z.shape[1], 3),
                                        dtype=np.float32)
-            self.generate_faces()
-            self.__meshdata.set_faces(self.__faces)
-            update_mesh = True
+            self.__faces = self._generate_faces()
+            new_vertices = True
+            update_faces = True
 
         # Copy x, y, z data into vertex array
-        if new_vertices or x is not None:
-            if x is None:
-                if self._x is None:
-                    x = np.arange(self._z.shape[0])
-                else:
-                    x = self._x
+        if new_vertices or x_is_new:
+            if not x_is_new and self._x is None:
+                x = np.arange(self._z.shape[0])
+            else:
+                x = self._x
             if x.ndim == 1:
                 x = x.reshape(len(x), 1)
             # Copy the 2D data into the appropriate slice
             self.__vertices[:, :, 0] = x
-            update_mesh = True
-
-        if new_vertices or y is not None:
-            if y is None:
-                if self._y is None:
-                    y = np.arange(self._z.shape[1])
-                else:
-                    y = self._y
+            update_vertices = True
 
+        if new_vertices or y_is_new:
+            if not y_is_new and self._y is None:
+                y = np.arange(self._z.shape[1])
+            else:
+                y = self._y
             if y.ndim == 1:
                 y = y.reshape(1, len(y))
-
             # Copy the 2D data into the appropriate slice
             self.__vertices[:, :, 1] = y
-            update_mesh = True
+            update_vertices = True
 
-        if new_vertices or z is not None:
+        if new_vertices or z_is_new:
             self.__vertices[..., 2] = self._z
-            update_mesh = True
+            update_vertices = True
+        return update_faces, update_vertices
 
-        if colors is not None:
-            self.__meshdata.set_vertex_colors(colors)
-            update_mesh = True
+    def _prepare_mesh_colors(self, colors):
+        if colors is None:
```

Comment:
@djhoese We could use `if colors is not None:` but this is just some style thing.

```python
if colors is not None:
    colors = np.asarray(colors)
    if colors.ndim == 3:
        # convert (width, height, 4) to (num_verts, 4)
        vert_shape = self.__vertices.shape
        num_vertices = vert_shape[0] * vert_shape[1]
        colors = colors.reshape(num_vertices, 3)
return colors
```

Reply:
Interesting. I disagree with this :wink: 

I like that the rest of the method isn't at an extra indentation level.

Reply:
Point taken! Let's move on :grinning: .